### PR TITLE
Fix compound matches with nodegroups

### LINKED ISSUE #52678: Compound targeting by nodegroup broken in 2019.2.0
### Description of Issue/Question

Compound targeting by nodegroup broken in 2019.2.0. Please see details below.

### Setup
```
root@master:/etc/salt/master.d# cat /etc/salt/master.d/nodegroups.conf 
nodegroups:
  testgroup: "G@virtual:VirtualBox"

root@master:/etc/salt/master.d# /etc/init.d/salt-master restart
[ ok ] Restarting salt-master (via systemctl): salt-master.service.
```

### Steps to Reproduce Issue
```
root@master:/etc/salt/master.d# salt -N testgroup test.ping
master:
    True
root@master:/etc/salt/master.d# salt -C 'G@virtual:VirtualBox' test.ping
master:
    True
root@master:/etc/salt/master.d# salt -C 'N@testgroup' test.ping
master:
    Minion did not return. [No response]
```

### Versions Report
```
root@master:/etc/salt/master.d# salt --versions-report
Salt Version:
           Salt: 2019.2.0
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.3
      docker-py: Not Installed
          gitdb: 2.0.0
      gitpython: 2.1.1
          ioflo: Not Installed
         Jinja2: 2.9.4
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.13 (default, Sep 26 2018, 18:42:22)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.4.3
            ZMQ: 4.2.1
 
System Versions:
           dist: debian 9.8 
         locale: UTF-8
        machine: x86_64
        release: 4.9.0-8-amd64
         system: Linux
        version: debian 9.8 
```

### PR DESCRIPTION
**Pull request by oeuftete**
### What does this PR do?

Fixes compound matches with nodegroups.

### What issues does this PR fix or reference?

Fixes #52678.

### Tests written?

Yes, some old commented tests were restored.

### Commits signed with GPG?

No

### CODE REVIEW

File: tests/integration/shell/test_matcher.py [hunk 1]
```diff
@@ -102,25 +102,24 @@ def test_compound_pillar(self):
...
-        # self.assertFalse(minion_in_returns('sub_minion', data))
-        # data = self.run_salt("-C 'N@multiline_nodegroup not @fakenodegroup not sub_minion' test.ping")
-        # self.assertTrue(minion_in_returns('minion', data))
-        # self.assertFalse(minion_in_returns('sub_minion', data))
+
+    def test_compound_nodegroup(self):
+        data = self.run_salt("-C 'N@multiline_nodegroup' test.ping")
+        self.assertTrue(minion_in_returns("minion", data))
+        self.assertTrue(minion_in_returns("sub_minion", data))
+        time.sleep(2)
```
Comment (oeuftete) [inline tests/integration/shell/test_matcher.py:114]: Does this need to be here?  I just kept what was commented out.  Does/did this make the test less flaky?

**Comment by oeuftete** [comment]
I need to investigate why the new test fails on Windows.

**Comment by oeuftete** [comment]
I need some help on how to fix the associated test here for Windows.

The functionality seems to work fine on Windows minions that are targeted from the master, or if I try something like `salt windows-minion match.compound 'N@somegroup'` (after adding the `nodegroups` config to the minion).  Is the Windows test effectively testing this change as if it was running from a Windows master?

**Review by DmitryKuzmenko — DISMISSED** [review]
LGTM

File: salt/client/__init__.py [hunk 2]
```diff
@@ -1677,15 +1677,37 @@ def _prep_pub(self, tgt, fun, arg, tgt_type, ret, jid, timeout, **kwargs):
         """
         Set up the payload_kwargs to be sent down to the master
         """
-        if tgt_type == "nodegroup":
-            if tgt not in self.opts["nodegroups"]:
+
+        def _resolve_nodegroup(ng):
```
Comment (dwoz) [inline salt/client/__init__.py:1681]: I'd prefer to see this method put at the module level, or as a class instance method. When methods are put inline like this it makes unit testing them much harder.

File: tests/integration/shell/test_matcher.py [hunk 3]
```diff
@@ -102,25 +102,23 @@ def test_compound_pillar(self):
 
     @skipIf(True, "This test is unreliable. Need to investigate why more deeply.")
     @flaky
-    def test_coumpound_pillar_pcre(self):
+    def test_compound_pillar_pcre(self):
         data = self.run_salt("-C 'J%@knights%^(Lancelot|Galahad)$' test.ping")
         self.assertTrue(minion_in_returns("minion", data))
         self.assertTrue(minion_in_returns("sub_minion", data))
-        # The multiline nodegroup tests are failing in develop.
```
Comment (dwoz) [inline tests/integration/shell/test_matcher.py:109]: If we are laving commented out code, I think we should create an issue to track the work that still needs to be done. We can assign the issue to you.

**Review by dwoz — CHANGES_REQUESTED** [review]

File: tests/integration/shell/test_matcher.py (hunk 3 above)
Comment (oeuftete) [reply tests/integration/shell/test_matcher.py:109]: It's the opposite here, I've reinstated some old commented-out tests.

File: salt/client/__init__.py (hunk 2 above)
Comment (oeuftete) [reply salt/client/__init__.py]: > I'd prefer to see this method put at the module level, or as a class instance method.

Done.

> When methods are put inline like this it makes unit testing them much harder.

Is there an implication here that you'd like to see a test that exercises this method directly?  Or is this only to make it more easily patchable if required by some other test?

Comment (waynew) [reply salt/client/__init__.py]: I don't think it's necessary to test that function, as long as it's being tested.

But in general, having functions at module level does allow for such things, and I feel certain I've seen cases in Salt where internal functions have grown beyond my screen size.

My personal approach is that unless there's a good reason to write an internal function, don't 😄

**Review by dwoz — APPROVED** [review]

**Pull request merged**